src/librustc_const/lib.rs

```diff
@@ -0,0 +1,88 @@
+//! This crates defines various helpers for const eval used by other crates
+//!
+//! For more information about how rustc works, see the [rustc dev guide].
+//!
+//! [rustc dev guide]: https://rustc-dev-guide.rust-lang.org/
+//!
+//! # Note
+//!
+//! This API is completely unstable and subject to change.
+
+#![doc(html_root_url = "https://doc.rust-lang.org/nightly/")]
+
+#[macro_use]
+extern crate log;
+
+use rustc::mir::interpret::LitToConstInput;
+use rustc::ty::subst::InternalSubsts;
+use rustc::ty::{self, Ty, TyCtxt};
+use rustc_hir as hir;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::ExprKind;
+
+/// Returns the `DefId` of the constant parameter that the provided expression is a path to.
+fn const_param_def_id(expr: &hir::Expr<'_>) -> Option<DefId> {
+    // Unwrap a block, so that e.g. `{ P }` is recognised as a parameter. Const arguments
+    // currently have to be wrapped in curly brackets, so it's necessary to special-case.
+    let expr = match &expr.kind {
+        ExprKind::Block(block, _) if block.stmts.is_empty() && block.expr.is_some() => {
+            block.expr.as_ref().unwrap()
+        }
+        _ => expr,
+    };
+
+    match &expr.kind {
+        ExprKind::Path(hir::QPath::Resolved(_, path)) => match path.res {
+            Res::Def(DefKind::ConstParam, did) => Some(did),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+pub fn ast_const_to_const<'tcx>(
```

Comment:
Wait, why is this its own crate?
This looks like it should be a query (taking a `DefId`, which a `hir::AnonConst` should always have) in `rustc_typeck`.

Reply:
feels a bit overkill to make it a query, all it does is intern a trivially created ty::Const

Reply:
I'd still be more comfortable with it if it were, idk. But also why is any of this necessary for repeat lengths, they can remain `Unevaluated` for as long as possible (you can use normalization to make it happen sooner without bothering to create anything other than `Unevaluated`).

Reply:
that's what I tried first. but that causes problems with parameters, because the parameters point to the surrounding function, causing cycles during mir building

Reply:
But they're not in the signature of the function, so there shouldn't be a cycle is my point.

Do you have one of those cycle errors? I'm intrigued now.

Reply:
Ah it's indeed not a cycle. It's an ICE.

```
error: internal compiler error: broken MIR in DefId(0:3 ~ issue_61336_1[317d]::f[0]) (_0 = [move _2; Const { ty: usize, val: Unevaluated(DefId(0:7 ~ issue_61336_1[317d]::f[0]::{{constant}}[1]), [T, Const { ty: usize, val: Param(N/#1) }], None) }]): bad assignment ([T; _] = [T; _]): NoSolution
  --> /home/oli/rustc/rust4/src/test/ui/const-generics/issues/issue-61336-1.rs:7:5
   |
LL |     [x; N]
   |   
```

at

https://github.com/rust-lang/rust/blob/42ce9b4e9054d63c8f352f994e2d5dae7d04130f/src/librustc_mir/borrow_check/type_check/mod.rs#L1437-L1440

Normalization doesn't convert unevaluated consts to `ConstKind::Param` where possible, and const comparisons of unevaluated constants compare the `DefId`, not the mir body.

Reply:
> Normalization doesn't convert unevaluated consts to `ConstKind::Param` where possible

Right, because we have nothing resembling symbolic miri. So we still need this hack.

If you don't want this to be a query... how about a constructor on `ty::Const`? That by default constructs a `ConstKind::Unevaluated`, but for local `DefId`s that point to `hir::AnonConst`s it does the hack.

Reply:
sgtm, the only reason I put it into a separate crate was that I didn't want to add stuff to the rustc crate

Reply:
done